clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/cache/OnDemandAware.groovy

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2016 Target, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.openstack.cache
+
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.netflix.spinnaker.cats.agent.CacheResult
+import com.netflix.spinnaker.cats.cache.CacheData
+import com.netflix.spinnaker.cats.cache.DefaultCacheData
+import com.netflix.spinnaker.cats.provider.ProviderCache
+import com.netflix.spinnaker.clouddriver.cache.OnDemandAgent.OnDemandResult
+import com.netflix.spinnaker.clouddriver.cache.OnDemandMetricsSupport
+import com.netflix.spinnaker.clouddriver.openstack.provider.view.MutableCacheData
+
+import java.util.concurrent.TimeUnit
+
+import static com.netflix.spinnaker.clouddriver.openstack.cache.Keys.Namespace.ON_DEMAND
+
+
+trait OnDemandAware {
+
+  /**
+   * Helper method to inspect onDemand 'toKeep' cache to see if the cacheData should be used
+   * instead of recreating it.
+   * @param cacheResultBuilder
+   * @param key
+   * @return
+   */
+  boolean shouldUseOnDemandData(CacheResultBuilder cacheResultBuilder, String key) {
+    CacheData cacheData = cacheResultBuilder.onDemand.toKeep[key]
+    cacheData ? cacheData.attributes.cacheTime >= cacheResultBuilder.startTime : false
+  }
+
+  /**
+   * Generically will move cached data from onDemand to resource (Server Group, Load Balancer, Security Group) namespace.
+   * @param objectMapper
+   * @param cacheResultBuilder
+   * @param serverGroupKey
+   */
+  void moveOnDemandDataToNamespace(ObjectMapper objectMapper, TypeReference<Map<String, List<MutableCacheData>>> typeReference, CacheResultBuilder cacheResultBuilder, String key) {
```

Comment:
This seems a bit confusing - you are mutating the parameters passed to this method. Might be more clear to return the cacheResultBuilder for clarity.